OpenBLAS 0.2.18 does not compile with PGI

Running into several issues with trying to compile on x86-64 based Linux system with PGI:

1) lapack/potf2/zpotf2_L.c fails with an 'Illegal lvalue' error:

`pgcc -c -O2 -DMAX_STACK_ALLOC=2048 -mp -tp p7-64 -DF_INTERFACE_PGI -fPIC -DDYNAMIC_ARCH -DSMP_SERVER -DUSE_OPENMP -DNO_WARMUP -DMAX_CPU_NUMBER=4 -DASMNAME=zpotf2_L -DASMFNAME=zpotf2_L_ -DNAME=zpotf2_L_ -DCNAME=zpotf2_L -DCHAR_NAME=\"zpotf2_L_\" -DCHAR_CNAME=\"zpotf2_L\" -DNO_AFFINITY -I../.. -DDOUBLE  -DCOMPLEX -DCOMPLEX -DDOUBLE zpotf2_L.c -o zpotf2_L.o`
`PGC-S-0058-Illegal lvalue (zpotf2_L.c: 71)`
`PGC/x86-64 Linux Rel Dev-r117179: compilation completed with severe errors`

This error also occurs with zpotf2_U.c.

Problem seems to be related to this construct, in common.h:

`#if defined(C_PGI) || defined(C_SUN)`
`#define CREAL(X)        (*((FLOAT *)&X + 0))`
`#define CIMAG(X)        (*((FLOAT *)&X + 1))`

If I change the code in zpotf2_L.c to use the standard C99 complex macros creal() and cimag(), the error goes away.

This problem also affects lapack/lauu2/zlauu2_L.c and lapack/lauu2/zlauu2_U.c.

2)  kernel/x86_64/cdot.c fails to compile:

`pgcc -c -O2 -DMAX_STACK_ALLOC=2048 -mp -tp p7-64 -DF_INTERFACE_PGI -fPIC -DDYNAMIC_ARCH -DSMP_SERVER -DUSE_OPENMP -DNO_WARMUP -DMAX_CPU_NUMBER=4 -DASMNAME=cdotc_k_SANDYBRIDGE -DASMFNAME=cdotc_k_SANDYBRIDGE_ -DNAME=cdotc_k_SANDYBRIDGE_ -DCNAME=cdotc_k_SANDYBRIDGE -DCHAR_NAME=\"cdotc_k_SANDYBRIDGE_\" -DCHAR_CNAME=\"cdotc_k_SANDYBRIDGE\" -DNO_AFFINITY -DTS=_SANDYBRIDGE -I.. -DBUILD_KERNEL -DTABLE_NAME=gotoblas_SANDYBRIDGE -UDOUBLE  -DCOMPLEX -DCOMPLEX -UDOUBLE -DCONJ ../kernel/x86_64/cdot.c -o cdotc_k_SANDYBRIDGE.o`
`PGC-S-0040-Illegal use of symbol, __real__ (../kernel/x86_64/cdot.c: 103)`
`PGC-W-0156-Type not specified, 'int' assumed (../kernel/x86_64/cdot.c: 103)`
`PGC-S-0040-Illegal use of symbol, __imag__ (../kernel/x86_64/cdot.c: 104)`
`PGC-W-0156-Type not specified, 'int' assumed (../kernel/x86_64/cdot.c: 104)`
`PGC-S-0043-Redefinition of symbol, result (../kernel/x86_64/cdot.c: 104)`
`PGC-S-0040-Illegal use of symbol, __real__ (../kernel/x86_64/cdot.c: 167)`
`PGC-W-0156-Type not specified, 'int' assumed (../kernel/x86_64/cdot.c: 167)`
`PGC-S-0043-Redefinition of symbol, result (../kernel/x86_64/cdot.c: 167)`
`PGC-S-0040-Illegal use of symbol, __imag__ (../kernel/x86_64/cdot.c: 168)`
`PGC-W-0156-Type not specified, 'int' assumed (../kernel/x86_64/cdot.c: 168)`
`PGC-S-0043-Redefinition of symbol, result (../kernel/x86_64/cdot.c: 168)`
`PGC/x86-64 Linux Rel Dev-r117182: compilation completed with severe errors`
`make[1]: *** [cdotc_k_SANDYBRIDGE.o] Error 2`

This is due to the use of a GNU-specific extension for C99 complex datatypes, which is not supported by PGI:

`__real__ result = 0.0 ;`
`__imag__ result = 0.0 ;`

One can fix this by changing the above as follows, to comply with the C99 standard:

`result = 0.0 + _Complex_I*0.0;`

This also affects kernel/x86_64/zdot.c.

Could you fix these issues for the next release?  Thanks!

--- a/common.h
+++ b/common.h
@@ -556,8 +556,13 @@ static void __inline blas_lock(volatile BLASULONG *address){
 #endif
 
 #if defined(C_PGI) || defined(C_SUN)
-#define CREAL(X)	(*((FLOAT *)&X + 0))
-#define CIMAG(X)	(*((FLOAT *)&X + 1))
+  #if defined(__STDC_IEC_559_COMPLEX__)
+     #define CREAL(X)   creal(X)
+     #define CIMAG(X)   cimag(X)
+  #else
+     #define CREAL(X)	(*((FLOAT *)&X + 0))
+     #define CIMAG(X)	(*((FLOAT *)&X + 1))
+  #endif
 #else
 #ifdef OPENBLAS_COMPLEX_STRUCT
 #define CREAL(Z)	((Z).real)

--- a/interface/zdot.c
+++ b/interface/zdot.c
@@ -160,6 +160,7 @@ OPENBLAS_COMPLEX_FLOAT CNAME(blasint n, FLOAT *x, blasint incx, FLOAT *y, blasin
 
   if (n <= 0) {
 #ifdef FORCE_USE_STACK
+    OPENBLAS_COMPLEX_FLOAT zero=OPENBLAS_MAKE_COMPLEX_FLOAT(0.0, 0.0);
     *result = zero;
 //	CREAL(*result) = 0.0;
 //	CIMAG(*result) = 0.0;

--- a/interface/zdot.c
+++ b/interface/zdot.c
@@ -160,9 +160,9 @@ OPENBLAS_COMPLEX_FLOAT CNAME(blasint n, FLOAT *x, blasint incx, FLOAT *y, blasin
 
   if (n <= 0) {
 #ifdef FORCE_USE_STACK
-    //*result = OPENBLAS_MAKE_COMPLEX_FLOAT(0.0, 0.0);
-	CREAL(*result) = 0.0;
-	CIMAG(*result) = 0.0;
+    *result = OPENBLAS_MAKE_COMPLEX_FLOAT(0.0, 0.0);
+//	CREAL(*result) = 0.0;
+//	CIMAG(*result) = 0.0;
     return;
 #else
     return zero;

--- a/interface/zdot.c
+++ b/interface/zdot.c
@@ -160,7 +160,7 @@ OPENBLAS_COMPLEX_FLOAT CNAME(blasint n, FLOAT *x, blasint incx, FLOAT *y, blasin
 
   if (n <= 0) {
 #ifdef FORCE_USE_STACK
-    *result = OPENBLAS_MAKE_COMPLEX_FLOAT(0.0, 0.0);
+    *result = zero;
 //	CREAL(*result) = 0.0;
 //	CIMAG(*result) = 0.0;
     return;

--- a/kernel/x86_64/cdot.c
+++ b/kernel/x86_64/cdot.c
@@ -100,8 +100,7 @@ FLOAT _Complex CNAME(BLASLONG n, FLOAT *x, BLASLONG inc_x, FLOAT *y, BLASLONG in
 
 	if ( n <= 0 ) 
 	{
-	        __real__ result = 0.0 ;
-        	__imag__ result = 0.0 ;
+	        result = OPENBLAS_MAKE_COMPLEX_FLOAT (0.0, 0.0) ;
 		return(result);
 
 	}
@@ -161,11 +160,13 @@ FLOAT _Complex CNAME(BLASLONG n, FLOAT *x, BLASLONG inc_x, FLOAT *y, BLASLONG in
 	}
 
 #if !defined(CONJ)
-	__real__ result = dot[0] - dot[1];
-	__imag__ result = dot[4] + dot[5];
+	        result = OPENBLAS_MAKE_COMPLEX_FLOAT (dot[0]-dot[1], dot[4]+dot[5]) ;
+//	CREAL(result) = dot[0] - dot[1];
+//	CIMAG(result) = dot[4] + dot[5];
 #else
-	__real__ result = dot[0] + dot[1];
-	__imag__ result = dot[4] - dot[5];
+	        result = OPENBLAS_MAKE_COMPLEX_FLOAT (dot[0]+dot[1], dot[4]-dot[5]) ;
+//	CREAL(result) = dot[0] + dot[1];
+//	CIMAG(result) = dot[4] - dot[5];
 
 #endif
 

--- a/kernel/x86_64/zdot.c
+++ b/kernel/x86_64/zdot.c
@@ -95,8 +95,9 @@ FLOAT _Complex CNAME(BLASLONG n, FLOAT *x, BLASLONG inc_x, FLOAT *y, BLASLONG in
 
 	if ( n <= 0 ) 
 	{
-	        __real__ result = 0.0 ;
-        	__imag__ result = 0.0 ;
+//	        CREAL(result) = 0.0 ;
+//        	CIMAG(result) = 0.0 ;
+		result=OPENBLAS_MAKE_COMPLEX_FLOAT(0.0,0.0);
 		return(result);
 
 	}
@@ -150,11 +151,13 @@ FLOAT _Complex CNAME(BLASLONG n, FLOAT *x, BLASLONG inc_x, FLOAT *y, BLASLONG in
 	}
 
 #if !defined(CONJ)
-	__real__ result = dot[0] - dot[1];
-	__imag__ result = dot[2] + dot[3];
+	result=OPENBLAS_MAKE_COMPLEX_FLOAT(dot[0]-dot[1],dot[2]+dot[3]);
+//	CREAL(result) = dot[0] - dot[1];
+//	CIMAG(result) = dot[2] + dot[3];
 #else
-	__real__ result = dot[0] + dot[1];
-	__imag__ result = dot[2] - dot[3];
+	result=OPENBLAS_MAKE_COMPLEX_FLOAT(dot[0]+dot[1],dot[2]-dot[3]);
+//	CREAL(result) = dot[0] + dot[1];
+//	CIMAG(result) = dot[2] - dot[3];
 
 #endif
 

